### ISSUE TITLE
Log to Metrics filter use container OOM Kill

### ISSUE THREAD

**Issue by 95jinhong** [fluent/fluent-bit#8593]
## Bug Report

**Describe the bug**

Fluent-bit containers leak memory (OOM) when applying Log to metric feature

**To Reproduce**
- Rubular link if applicable: https://docs.fluentbit.io/manual/pipeline/filters/log_to_metrics
- Example log message if applicable:

```
## pod state
NAME                   READY   STATUS    RESTARTS        AGE
ddocdoc-metric-bl7ph   1/1     Running   4 (4h8m ago)    47h
ddocdoc-metric-g4sdp   1/1     Running   0               47h
ddocdoc-metric-nq2xp   1/1     Running   4 (8h ago)      47h
ddocdoc-metric-qggkh   1/1     Running   4 (7h58m ago)   47h

## pod describe
    State:          Running
      Started:      Sun, 17 Mar 2024 12:56:37 +0900
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
      Started:      Sat, 16 Mar 2024 11:43:43 +0900
      Finished:     Sun, 17 Mar 2024 12:56:35 +0900
    Ready:          True
    Restart Count:  4
    Limits:
      cpu:     2
      memory:  6Gi

## fluent-bit pod log
# not error, flush log
[2024/03/17 08:02:15] [debug] [out flush] cb_destroy coro_id=4924
[2024/03/17 08:02:16] [debug] [out flush] cb_destroy coro_id=4925
[2024/03/17 08:02:16] [debug] [out flush] cb_destroy coro_id=4926
[2024/03/17 08:07:32] [debug] [input:tail:tail.0] inode=104256582, /var/log/containers/dd-reception-hub-consumer-58bfbc854b-lvh2f_reception_server-249c9f46e2bca5ee14d89f398cbf12c515aac899ed28e4432a52c38746f0e70b.log, events: IN_MODIFY
[2024/03/17 08:07:32] [debug] [input chunk] update output instances with new chunk size diff=2681471, records=0, input=emitter.2
[2024/03/17 08:07:32] [debug] [input chunk] update output instances with new chunk size diff=2681747, records=0, input=emitter.2
[2024/03/17 08:07:32] [debug] [input chunk] update output instances with new chunk size diff=10644, records=5, input=tail.0
```
```
## fluent-bit configmap
    [INPUT]
        Name               tail
        Tag                kube.reception.*
        Path               /var/log/containers/dd-reception-hub-consumer*.log
        multiline.parser  docker, cri
        Refresh_Interval   10
        Skip_Long_Lines    Off

    [FILTER]
        Name                parser
        Match               kube.reception.*
        Key_Name            log
        Parser              json
        Reserve_Data        On

    [FILTER]
        name               log_to_metrics
        match              kube.reception.*
        tag                test_metric
        metric_mode        gauge
        metric_name        current_hospital_jobs
        metric_description  This metric shows the current_hospital_jobs
        value_field        count
        kubernetes_mode    on
        regex              message .*CHANGE_HOSPITAL_JOBS*
        label_field        hospitalId
        label_field        operation
        label_field        status
        label_field        logCreatedAt

    [OUTPUT]
        name               prometheus_exporter
        match              *
        host               0.0.0.0
        port               2021

## log to metric example
log_metric_gauge_current_hospital_jobs{namespace_name="reception",pod_name="dd-reception-hub-consumer-58bfbc854b-7fbbf",container_name="server",pod_id="9789c027-8545-4e5d-b9c9-a4ea804d63c9",hospitalId="1234567890",operation="NEW_RECEIPT",status="REQUESTED",logCreatedAt="1710633626712"} 
```

- Steps to reproduce the problem:
1.  user is logged at a certain time of day with an influx of traffic, fluent-bit pod memory usage will go up at that time.
2. I think.. GC in fluent-bit doesn't seem to be working.
(currently my flush is set to 1 second)
3. During off-peak hours, changing logs to metrics does not increase memory usage.

**Expected behavior**

1. The number of logs in a one-hour period at the time of the memory utilization spike is 48000
![스크린샷 2024-03-17 17 20 41](https://github.com/fluent/fluent-bit/assets/43333982/033a34cd-1e08-4aff-872e-86069413d54a)

2. I noticed an increase in memory usage from 08:30 to 09:30.
![스크린샷 2024-03-17 17 18 46](https://github.com/fluent/fluent-bit/assets/43333982/ad0e9a6f-cc4c-4b3d-becd-a40eb54c3fc0)

**Screenshots**

**Your Environment**

* Version used: 2.2
* Configuration: Helm 
* Environment name and version (e.g. Kubernetes? What version?): AWS EKS 1.25 
* Server type and version:  Amazon Linux
* Operating System and version: Prometheus Stack v0.68.0
* Filters and plugins: log to metrics

**Additional context**

I think I have a memory leak, as the OOM only occurs when user traffic increases. Any idea what options I should set?

Also, since the information changed by log to metric is stored in memory, when fluent-bit gets the data from prometheus, when does it disappear from memory? 
I have set flush to 1 second, but I wonder if the flush option gc is working through the flush option?

**Comment by lecaros** [comment]
Hi there. Are you consuming the data from the exporter? Does it reproduce is you use another output? (stdout, null, http...)

**Comment by 95jinhong** [comment]
> Hi there. Are you consuming the data from the exporter? Does it reproduce is you use another output? (stdout, null, http...)

Hi, @lecaros 
We're currently exposing through the prometheus_exporter, which is shown as an example in the official documentation, and registering as a servicemonitor in prometheus to get the metrics.

I haven't tried other outputs, but when the traffic is low, the metrics are generated normally, but when the traffic is high and the logs are large, Fluent-bit causes a memory leak.
(I'll test with other outputs and let you know.)
Looking at this, the output, prometheus_exporter, seems to be working fine, but the GC is not working properly.

**Comment by drbugfinder-work** [comment]
Maybe related to this: https://github.com/fluent/fluent-bit/pull/8422

**Comment by 95jinhong** [comment]
> Maybe related to this: #8422

Thanks for letting us know.
I'll try it and let you know when it's available in 3.0.0~!

I'm testing it by adjusting various options in Fluent-bit, but still causing OOM whenever there's a lot of logs generated at once due to instant traffic...

**Comment by 95jinhong** [comment]
Hi, @drbugfinder-work 
I bumped it up to 3.0.0 and it seems to have fixed the Fluent-bit OOM issue. I will add the plugin you mentioned and test it and share it with you.
